src/VisualStudio/Core/Def/Implementation/DiagnosticCache/VisualStudioDiagnosticCacheService_ErrorList.cs

```diff
@@ -0,0 +1,153 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Threading;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Roslyn.Utilities;
+
+namespace Microsoft.VisualStudio.LanguageServices.Implementation.DiagnosticCache
+{
+    internal sealed partial class VisualStudioDiagnosticCacheService
+    {
+        private const string DiagnosticsUpdatedEventName = "CachedDiagnosticsUpdated";
+
+        private readonly object _gate = new();
+
+        // This is null from when live analysis started until a new solution is opened,
+        // which means no cached diagnostic update is allowed.
+        private string? _currentSolutionPath;
+
+        // Keeps track of all documents we pushed cached diagnostics for.
+        private readonly Dictionary<DocumentId, DiagnosticsUpdatedArgs> _requestedDocuments = new();
+
+        // Keeps track of documents encountered for live analysis.
+        private readonly HashSet<DocumentId> _analyzedDocuments = new();
+
+        private readonly TaskQueue _errorListUpdateQueue;
+
+        public event EventHandler<DiagnosticsUpdatedArgs>? CachedDiagnosticsUpdated;
+
+        private void TryUpdateDiagnosticsLoadedFromCache(Document document, ImmutableArray<DiagnosticData> diagnostics)
+        {
+            Debug.Assert(!diagnostics.IsDefaultOrEmpty);
+
+            var solutionPath = document.Project.Solution.FilePath;
+            if (solutionPath == null)
+            {
+                return;
+            }
+
+            lock (_gate)
+            {
+                // This check also ensures no update being pushed if _currentSolutionPath is null,
+                // which means live analysis already started.
+                if (solutionPath == _currentSolutionPath && !_requestedDocuments.ContainsKey(document.Id))
+                {
+                    var eventArg = DiagnosticsUpdatedArgs.DiagnosticsCreated(
+                            new CachedDiagnosticsUpdateArgsId(document.Id), document.Project.Solution.Workspace, document.Project.Solution, document.Project.Id, document.Id, diagnostics);
+                    _requestedDocuments.Add(document.Id, eventArg);
+                    _ = _errorListUpdateQueue.ScheduleTask(DiagnosticsUpdatedEventName, () =>
+                    {
+                        if (_diagnosticService is DiagnosticService s)
+                        {
+#pragma warning disable RS0043 // Do not call 'GetTestAccessor()'
+                            s.GetTestAccessor().EventListenerTracker.EnsureEventListener(_workspace, s);
```

Comment:
@mavasani ErrorList lazily listens to diagnostic service, so it's only started when there's any update from the service. What's the best way to start errorlist before we start pushing cached diagnostics to it?

Reply:
@genlu Live diagnostics table source is currently hooked up to be in stable state when solution crawler is not running or has not yet started. You can see the implementation here: https://github.com/dotnet/roslyn/blob/a4bdff76ec342a445d6c59025df96f55688cb8d5/src/VisualStudio/Core/Def/Implementation/TableDataSource/VisualStudioBaseDiagnosticListTable.LiveTableDataSource.cs#L83-L92

IsSolutionCrawlerRunning is implemented here:
https://github.com/dotnet/roslyn/blob/a4bdff76ec342a445d6c59025df96f55688cb8d5/src/VisualStudio/Core/Def/Implementation/TableDataSource/AbstractRoslynTableDataSource.cs#L33-L83

Basically, if neither the build nor solution crawler are running, the live diagnostic source will always mark error list stable, which prevents the underlying error list to try to fetch updated diagnostics. You need to change the above logic to also account for cache diagnostics availability.